File: 이준영/백준/Gold/15686. 치킨 배달/치킨 배달.py
N, M = map(int, input().split())
arr = [list(map(int, input().split())) for _ in range(N)]
home = []
chicken = []
for i in range(N):
    for j in range(N):
        if arr[i][j] == 1:
            home.append((i, j))
        elif arr[i][j] == 2:
            chicken.append((i, j))
h_len = len(home)
c_len = len(chicken)
dis = [[0] * c_len for _ in range(h_len)]
for i in range(h_len):
    for j in range(c_len):
        dis[i][j] = (abs(home[i][0] - chicken[j][0]) + abs(home[i][1] - chicken[j][1]), j)  # (거리, 치킨집인덱스)
    dis[i].sort()

mn = 100000
for l in range(1 << c_len):  # 비트마스킹
    cnt = 0
    check = 0
    for i in range(c_len):
        if (l & (1 << i)) != 0:
            check += 1
    if check != M:           # 고른 치킨집이 M개 일 때만 체크
        continue
    for i in range(h_len):
        for j in range(c_len):
            if (l & (1 << dis[i][j][1])) != 0:  # 미리 정렬을 했기에 조건 만족하는 치킨집있으면 바로 반영
                cnt = cnt + dis[i][j][0]
                break
    mn = min(mn, cnt)
print(mn)

File: 이준영/백준/Gold/15686. 치킨 배달/치킨 배달.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.Map;
import java.util.StringTokenizer;

public class Main {
	static int minimum = 1000000;
	static int M=0;
	static int len2=0;
	static int len1=0;
	static Map<Integer, Boolean> map = new HashMap<>();
	static ArrayList<int[]>[] distance;
	static ArrayList<int[]> chicken;
	public static void main(String[] args) throws Exception{
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = new StringTokenizer(br.readLine());
		int N = Integer.parseInt(st.nextToken());
		M = Integer.parseInt(st.nextToken());
		int[][] arr = new int[N][N];
		ArrayList<int[]> home = new ArrayList<>();
		chicken = new ArrayList<>();
		map = new HashMap<>();
		for(int i=0;i<N;i++) {
			st = new StringTokenizer(br.readLine());
			for(int j=0;j<N;j++) {
				int a = Integer.parseInt(st.nextToken());
				arr[i][j]=a;
				if(a==1) {
					home.add(new int[] {i,j});
				} else if(a==2) {
					chicken.add(new int[] {i,j});
					map.put(51*i+j, false);
				}
			}
		}
		len1 = home.size();
		len2 = chicken.size();
		distance = new ArrayList[len1];
 		for(int i=0;i<len1;i++) {
			distance[i]=new ArrayList<>();
			int[] tmp=home.get(i);
			for(int j=0;j<len2;j++) {
				int[] tmp2 = chicken.get(j);
				distance[i].add(new int[] {Math.abs(tmp[0]-tmp2[0])+Math.abs(tmp[1]-tmp2[1]),tmp2[0],tmp2[1]});
			}
			Collections.sort(distance[i], (a,b)->a[0]-b[0]);
		}
 		btk(0,0);
 		System.out.println(minimum);	
	}
	static void btk(int cur, int count) {
		if(count==M) {
			check();
			return;
		}
		if(cur==len2) {
			return;
		}
		btk(cur+1,count);
		int[] tmp=chicken.get(cur);
		map.put(tmp[0]*51+tmp[1], true);
		btk(cur+1,count+1);
		map.put(tmp[0]*51+tmp[1], false);
	}
	static void check() {
		int sum=0;
		for(int i=0;i<len1;i++) {
			for(int j=0;j<len2;j++) {
				int[] tmp = distance[i].get(j);
				if(map.get(tmp[1]*51+tmp[2])) {
					sum+=tmp[0];
					break;
				}
			}
		}
		if (sum<minimum) {
			minimum=sum;
		return;
		}
	}
}